File: Stacks/StackArray.java
public class StackArray <T> {
    private int size;
    private int top;
    private T s[];
    
    // Default constructor
    public StackArray(){
        this.size = 10;
        this.top = -1;
        this.s = (T[]) new Object[10];
    }
    
    // Parameterized Constructor
    public StackArray(int size){
        if(size < 1)
            throw new IllegalArgumentException("Minimum stack size is 1!");
        this.size = size;
        this.top = -1;
        this.s = (T[]) new Object[size];
    }

    public boolean isFull(){
        return this.top == this.size - 1; 
    }
    
    public boolean isEmpty(){
        return this.top == -1;
    }

    public boolean push(T val){
        if(this.isFull()){
            System.out.println("Stack overflow!");
            return false;
        }
        
        this.top++;
        this.s[this.top] = val;
        return true;
    }
    
    public T pop(){
        if(isEmpty()){
            System.out.println("Stack is empty! Nothing to pop!");
            return null;
        }
        
        return this.s[this.top--];
    }
    
    public T peek(int position){
        if(isEmpty()){
            System.out.println("Stack is empty! Nothing to peek!");
            return null;
        }
        
        int index = this.top - position + 1;
        
        if(index < 0 || index > this.top){
            System.out.println("Invalid position!");
            return null;
        }
        
        return this.s[index];
    }
    
    public T stackTop(){
        if(isEmpty()){
            System.out.println("Stack is empty!");
            return null;
        }
        
        return this.s[this.top];
    }
    
    public int size(){
        return this.top + 1;
    }
}

File: Stacks/StackArray.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct {
    int size;
    int top;
    int* s;
} Stack;

void create(Stack* st){
    printf("Enter stack size: ");
    scanf("%d", &st->size);
    st->top = -1;
    st->s = malloc(st->size * sizeof(int));
}

void display(Stack st){
    printf("Stack:\n");
    for (int i = st.top; i >= 0; i--){
        printf("%d\n", st.s[i]);
    }
}

//Stack funcs
bool isFull(Stack st){        
    return (st.top == st.size - 1);
}

bool isEmpty(Stack st){
    return (st.top == -1);
}

bool push(Stack *st, int val){
    if(isFull(*st)){
        printf("Stack Overflow!\n");
        return false;
    }
    
    st->s[st->top + 1] = val;
    st->top++;    
    return true;
}

int pop(Stack *st){
    if(isEmpty(*st)){
        printf("Stack is empty\n");
        return -1;
    }
    
    
    st->top--;
    return st->s[st->top + 1];
}

int peek(Stack st, int pos){
    if(isEmpty(st)){
        printf("Stack is empty\n");
        return -1;
    }
    
    int index = st.top - pos + 1;
    
    return st.s[index];
}

int stack_top(Stack st){
    if(!isEmpty(st))
        return st.s[st.top];
}

int main(void){
    Stack st;
    create(&st);
    
    push(&st, 10);
    push(&st, 20);
    push(&st, 30);
    display(st);
    printf("\n");
    
    // printf("%d \n", pop(&st));
    // printf("%d \n", pop(&st));
    // printf("%d \n", pop(&st));
    // printf("%d \n", pop(&st));
    
    // display(st);
    printf("%d \n", peek(st, 3));
    printf("%d \n", peek(st, 2));
    printf("%d \n", peek(st, 1));
    
    printf("Stack top: %d\n", stack_top(st));
    
    printf("Is Empty? %s\n", isEmpty(st));
    printf("Is Full? %s\n", isFull(st));
    
    printf("Empty spaces: %d\n", st.size - st.top - 1);
    return 0;
}